File: exercicios/kth _largest_element_in_an_array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        return self.quickselect(nums, 0, len(nums) - 1, len(nums) - k)

    def quickselect(self, nums: List[int], left: int, right: int, k_smallest: int) -> int:
        if left == right:
            return nums[left]

        pivot_index = self.partition(nums, left, right)
        if k_smallest == pivot_index:
            return nums[k_smallest]
        elif k_smallest < pivot_index:
            return self.quickselect(nums, left, pivot_index - 1, k_smallest)
        else:
            return self.quickselect(nums, pivot_index + 1, right, k_smallest)

    def partition(self, nums: List[int], left: int, right: int) -> int:
        pivot = nums[right]
        i = left - 1

        for j in range(left, right):
            if nums[j] <= pivot:
                i += 1
                nums[i], nums[j] = nums[j], nums[i]

        nums[i + 1], nums[right] = nums[right], nums[i + 1]
        return i + 1

File: exercicios/kth_largest_element_in_an_array.c
int partition(int* v,int leftIndex, int rightIndex) {

    int valueToCompare = v[rightIndex];
    int j = leftIndex;

    for(int k = leftIndex; k< rightIndex ; k++)
        if(v[k]<=valueToCompare)
        {

            int temporaryValue = v[k];
            v[k] = v[j];
            v[j] = temporaryValue;
            j++;

        }

    int temporaryValue = v[j];
    v[j] = v[rightIndex];
    v[rightIndex] = temporaryValue;

    return j;
}

void quickSelect(int* v, int valueToSort, int leftIndex, int rightIndex) {

	int j = partition(v, leftIndex, rightIndex);

	if(j == valueToSort) return;
	if(j < valueToSort)
		quickSelect(v, valueToSort, j+1, rightIndex);
	else
		quickSelect(v, valueToSort, leftIndex, j-1);

}

int findKthLargest(int* nums, int numsSize, int k) {
    k = numsSize - k;
    quickSelect(nums, k, 0, numsSize - 1);
    return nums[k];
}
